### PR TITLE
Deprecate the Controller::addImageToTemplate() method

### PR DESCRIPTION
**Pull request by m-vo**
| Q                | A
| -----------------| ---
| Fixed issues     | -
| Docs PR or issue | -

This deprecates `Controller::addImageToTemplate` and rewires all usages. What I really like is that you can now see what the element's/module's figure is built from. Like: Is there a resize happening? From what data? Lightbox? Overwritten mtadata? ....

**Example**

before:
```php
// Add an image
if ($this->addImage && $this->singleSRC != '')
{
    $objModel = FilesModel::findByUuid($this->singleSRC);

    if ($objModel !== null && is_file(System::getContainer()->getParameter('kernel.project_dir') . '/' . $objModel->path))
    {
        $this->singleSRC = $objModel->path;
        $this->addImageToTemplate($this->Template, $this->arrData, null, null, $objModel);
    }
}
```

after:
```php
 // Add an image
if ($this->addImage && null !== ($figureBuilder = $this->getFigureBuilderIfResourceExists($this->singleSRC)))
{
    $figureBuilder
        ->setSize($this->size)
        ->setMetaData($this->objModel->getOverwriteMetaData())
        ->enableLightbox($this->fullsize)
        ->build()
        ->applyLegacyTemplateData($this->Template, $this->imagemargin, $this->floating);
}
```

This PR needs to be rebased once #2054 + #2060 are merged upstream and depends on #2071 for BC compatibility.

### CODE REVIEW

**Comment by m-vo** [comment]
In the meantime lots of conflicts happened. Do we want these changes?

I don't want to waste my time resolving + adjusting if not. :relaxed:

**Comment by ausi** [comment]
> Do we want these changes?

👍 I’m definetly in favor of deprecating `Controller::addImageToTemplate()`

**Comment by fritzmg** [comment]
Same here 👍 

We have a mumble call tomorrow, we could at least discuss it briefly then, if no other feedback arrives here ;)

File: core-bundle/src/Image/Studio/LegacyFigureBuilderTrait.php [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+use Contao\CoreBundle\Exception\InvalidResourceException;
+use Contao\System;
+
+/**
+ * This trait is intended to be used in legacy content elements and modules
+ * where dependency injection isn't available and missing images are silently
+ * ignored for BC reasons.
+ */
+trait LegacyFigureBuilderTrait
```
Comment (m-vo) [inline core-bundle/src/Image/Studio/LegacyFigureBuilderTrait.php:25]: todo: make `@internal`

**Comment by m-vo** [comment]
PR is rebased + should be ready to go now. :tada: 

I tried to stick to closely to the old behavior but as there are no tests for the legacy code it might be helpful to test these changes in a real application as well.

File: calendar-bundle/src/Resources/contao/modules/ModuleEventReader.php [hunk 2]
```diff
@@ -287,29 +290,14 @@ protected function compile()
...
-					{
-						$arrEvent['size'] = $this->imgSize;
-					}
-				}
-
-				$arrEvent['singleSRC'] = $objModel->path;
-				$this->addImageToTemplate($objTemplate, $arrEvent, null, null, $objModel);
-			}
+			$figureBuilder
+				->setSize($objEvent->size ?: $this->imgSize ?: null)
```
Comment (ausi) [inline calendar-bundle/src/Resources/contao/modules/ModuleEventReader.php]: This now prefers `$objEvent->size` whereas the old code preferred `$this->imgSize` if it was set to a non-empty size.

File: calendar-bundle/src/Resources/contao/modules/ModuleEventlist.php [hunk 3]
```diff
@@ -358,32 +361,28 @@ protected function compile()
...
-					{
-						$event['size'] = $imgSize;
-					}
+				/** @var CalendarEventsModel $eventModel */
+				$eventModel = CalendarEventsModel::findByPk($event['id']);
 
-					$event['singleSRC'] = $objModel->path;
-					$this->addImageToTemplate($objTemplate, $event, null, null, $objModel);
+				$figure = $figureBuilder
+					->setSize($eventModel->size ?: $this->imgSize ?: null)
```
Comment (ausi) [inline calendar-bundle/src/Resources/contao/modules/ModuleEventlist.php]: Same here.

**Review by ausi — CHANGES_REQUESTED** [review]

File: calendar-bundle/src/Resources/contao/modules/ModuleEventReader.php (hunk 2 above)
Comment (m-vo) [reply calendar-bundle/src/Resources/contao/modules/ModuleEventReader.php]: Fixed in 1db55b3d80534f519efddf28902881709787bf51.

File: calendar-bundle/src/Resources/contao/modules/ModuleEventlist.php (hunk 3 above)
Comment (m-vo) [reply calendar-bundle/src/Resources/contao/modules/ModuleEventlist.php]: Fixed in 1db55b3d80534f519efddf28902881709787bf51.

File: calendar-bundle/src/Resources/contao/modules/ModuleEventlist.php [hunk 4]
```diff
@@ -358,32 +361,28 @@ protected function compile()
...
-					{
-						$event['size'] = $imgSize;
-					}
+				/** @var CalendarEventsModel $eventModel */
+				$eventModel = CalendarEventsModel::findByPk($event['id']);
 
-					$event['singleSRC'] = $objModel->path;
-					$this->addImageToTemplate($objTemplate, $event, null, null, $objModel);
+				$figure = $figureBuilder
+					->setSize($this->imgSize ?: $eventModel->size ?: null)
```
Comment (ausi) [inline calendar-bundle/src/Resources/contao/modules/ModuleEventlist.php]:
```suggestion
					->setSize($imgSize ?: $eventModel->size ?: null)
```

File: calendar-bundle/src/Resources/contao/modules/ModuleEventReader.php [hunk 5]
```diff
@@ -287,29 +290,14 @@ protected function compile()
...
-					{
-						$arrEvent['size'] = $this->imgSize;
-					}
-				}
-
-				$arrEvent['singleSRC'] = $objModel->path;
-				$this->addImageToTemplate($objTemplate, $arrEvent, null, null, $objModel);
-			}
+			$figureBuilder
+				->setSize($this->imgSize ?: $objEvent->size ?: null)
```
Comment (ausi) [inline calendar-bundle/src/Resources/contao/modules/ModuleEventReader.php]: This would now use `$this->imgSize` even if it is set to an “empty” value like `a:3:{i:0;i:0;i:1;i:0;i:2;s:3:"box";}`. We should probably use an $imgSize variable like *ModuleEventlist.php* does in line 284.

**Review by ausi — CHANGES_REQUESTED** [review]

File: calendar-bundle/src/Resources/contao/modules/ModuleEventReader.php (hunk 5 above)
Comment (m-vo) [reply calendar-bundle/src/Resources/contao/modules/ModuleEventReader.php]: see 93f7e0376750f50ca7999490baea7f917de4cc87

File: core-bundle/src/Resources/contao/elements/ContentImage.php [hunk 6]
```diff
@@ -36,6 +40,8 @@ class ContentImage extends ContentElement
 	 */
 	public function generate()
 	{
+		// Note: These file operations are only here for BC reasons and do not
+		//       serve any functional purpose anymore.
```
Comment (ausi) [inline core-bundle/src/Resources/contao/elements/ContentImage.php]: I’m a little bit confused. Does’t this method set up `$this->objFilesModel` which is used afterwards in line 68 by `$this->getFigureBuilderIfResourceExists($this->objFilesModel);`?

File: core-bundle/src/Resources/contao/elements/ContentText.php [hunk 7]
```diff
@@ -42,15 +46,14 @@ protected function compile()
...
-
-			if ($objModel !== null && is_file(System::getContainer()->getParameter('kernel.project_dir') . '/' . $objModel->path))
-			{
-				$this->singleSRC = $objModel->path;
-				$this->addImageToTemplate($this->Template, $this->arrData, null, null, $objModel);
-			}
+			$figureBuilder
+				->setSize($this->size)
+				->setMetaData($this->objModel->getOverwriteMetaData())
+				->enableLightbox()
```
Comment (ausi) [inline core-bundle/src/Resources/contao/elements/ContentText.php]: Shouldn’t this be
```suggestion
				->enableLightbox($this->fullsize)
```
?

File: faq-bundle/src/Resources/contao/modules/ModuleFaqPage.php [hunk 8]
```diff
@@ -101,19 +103,17 @@ protected function compile()
...
-				}
+				$figureBuilder
+					->setSize($objFaq->size)
+					->setMetaData($objFaq->getOverwriteMetaData())
+					->setLightboxGroupIdentifier('lightbox[' . substr(md5('mod_faqpage_' . $objFaq->id), 0, 6) . ']')
+					->enableLightbox($objFaq->fullsize)
+					->build()
+					->applyLegacyTemplateData($objTemp, $objFaq->imagemargin, $objFaq->floating);
+
+				$this->Template->useImage = true;
```
Comment (ausi) [inline faq-bundle/src/Resources/contao/modules/ModuleFaqPage.php]: This line seems unrelated here. Maybe a copy & paste error?

File: faq-bundle/src/Resources/contao/modules/ModuleFaqReader.php [hunk 9]
```diff
@@ -124,18 +127,16 @@ protected function compile()
...
-				$this->addImageToTemplate($this->Template, $arrFaq, null, null, $objModel);
-			}
+			$figureBuilder
+				->setSize($objFaq->size)
+				->setMetaData($objFaq->getOverwriteMetaData())
+				->enableLightbox($objFaq->fullsize)
+				->build()
+				->applyLegacyTemplateData($this->Template, $objFaq->imagemargin, $objFaq->floating);
+
+			$this->Template->useImage = true;
```
Comment (ausi) [inline faq-bundle/src/Resources/contao/modules/ModuleFaqReader.php]: Same here?

File: news-bundle/src/Resources/contao/modules/ModuleNews.php [hunk 10]
```diff
@@ -159,49 +162,34 @@ protected function parseArticle($objArticle, $blnAddArchive=false, $strClass='',
...
 
-				// Link to the news article if no image link has been defined (see #30)
-				if (!$objTemplate->fullsize && !$objTemplate->imageUrl)
-				{
-					// Unset the image title attribute
-					$picture = $objTemplate->picture;
-					unset($picture['title']);
-					$objTemplate->picture = $picture;
+			$figure = $figureBuilder
+				->setSize($objArticle->size)
```
Comment (ausi) [inline news-bundle/src/Resources/contao/modules/ModuleNews.php]: The `$this->imgSize` override seems to be missing here.

**Review by ausi — CHANGES_REQUESTED** [review]
I now reviewed the whole PR. Apart from my comments this looks very good IMO 🎉

But I think it would make sense in this case to test all the elements/modules in an actual Contao setup ☺️.

File: core-bundle/src/Resources/contao/elements/ContentImage.php (hunk 6 above)
Comment (m-vo) [reply core-bundle/src/Resources/contao/elements/ContentImage.php]: Ah yes, this comment is misleading. I initially had directly used `$this->singleSRC` as the resource for the `FigureBuilder` which made this whole method obsolete. But `singleSRC` being a path instead of a UUID and `objFilesModel` being actually set could be something people are relying on and I'm not sure we can drop for BC reasons. Wdyt? 

If we cannot drop it, we should imo just remove the comment.

File: core-bundle/src/Resources/contao/elements/ContentText.php (hunk 7 above)
Comment (m-vo) [reply core-bundle/src/Resources/contao/elements/ContentText.php]: :+1: Yes, looks like I missed that one.

File: faq-bundle/src/Resources/contao/modules/ModuleFaqPage.php (hunk 8 above)
Comment (m-vo) [reply faq-bundle/src/Resources/contao/modules/ModuleFaqPage.php]: I was wondering about this the last time I visited the place but was sure this must have a reason... Turns out I cannot find one. :smile: 

Removed in d70f0466606df10fdfff2f1da509aef761c37d4c

File: faq-bundle/src/Resources/contao/modules/ModuleFaqReader.php (hunk 9 above)
Comment (m-vo) [reply faq-bundle/src/Resources/contao/modules/ModuleFaqReader.php]: fixed in d70f0466606df10fdfff2f1da509aef761c37d4c

File: news-bundle/src/Resources/contao/modules/ModuleNews.php (hunk 10 above)
Comment (m-vo) [reply news-bundle/src/Resources/contao/modules/ModuleNews.php]: Fixed in 762b0f65877d986a37bfafcfafb4ac6195ca5b2e

File: core-bundle/src/Resources/contao/elements/ContentImage.php (hunk 6 above)
Comment (ausi) [reply core-bundle/src/Resources/contao/elements/ContentImage.php]: I think we should just remove the comment.

**Review by ausi — DISMISSED** [review]

File: core-bundle/src/Resources/contao/elements/ContentImage.php (hunk 6 above)
Comment (m-vo) [reply core-bundle/src/Resources/contao/elements/ContentImage.php]: :+1: removed in 11a205ea8b839433445719f1f1e200de8da21ced

**Review by ausi — DISMISSED** [review]

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @m-vo.
